### ISSUE TITLE
Cmd escaping is causing argument parse errors

### ISSUE THREAD

**Issue by weshaggard** [dotnet/cli#181]
https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/Command.cs#L64
The escaping of quotes appear to be unnecessary in some scenarios.

```
> .\dotnet -v publish ..\test -f dnxcore50 -r win7-x64
Publishing test for DNXCore,Version=v5.0/win7-x64
> C:\Windows\system32\cmd.exe /C "dotnet-compile --framework \"DNXCore,Version=v5.0\" --output \"C:\temp\dotnet_test\tes
t\bin\Debug\dnxcore50\win7-x64\" --configuration \"Debug\" \"C:\temp\dotnet_test\test\""
System.ArgumentNullException: Value cannot be null.
Parameter name: frameworkVersion
   at NuGet.Frameworks.NuGetFramework..ctor(String frameworkIdentifier, Version frameworkVersion, String frameworkProfil
e)
   at NuGet.Frameworks.NuGetFramework.ParseFrameworkName(String frameworkName, IFrameworkNameProvider mappings)
   at NuGet.Frameworks.NuGetFramework.Parse(String folderName, IFrameworkNameProvider mappings)
   at NuGet.Frameworks.NuGetFramework.Parse(String folderName)
   at Microsoft.DotNet.Tools.Compiler.Program.<>c__DisplayClass0_1.<Main>b__1(String f) in D:\git\cli\src\Microsoft.DotN
et.Tools.Compiler\Program.cs:line 54
   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()
   at Microsoft.DotNet.Tools.Compiler.Program.<>c__DisplayClass0_0.<Main>b__0() in D:\git\cli\src\Microsoft.DotNet.Tools
.Compiler\Program.cs:line 56
   at Microsoft.Dnx.Runtime.Common.CommandLine.CommandLineApplication.Execute(String[] args) in C:\Users\Wes\.dnx\packag
es\Microsoft.Extensions.CommandLineUtils.Sources\1.0.0-rc2-16169\shared\CommandLineApplication.cs:line 263
   at Microsoft.DotNet.Tools.Compiler.Program.Main(String[] args) in D:\git\cli\src\Microsoft.DotNet.Tools.Compiler\Prog
ram.cs:line 70
< C:\Windows\system32\cmd.exe /C "dotnet-compile --framework \"DNXCore,Version=v5.0\" --output \"C:\temp\dotnet_test\tes
t\bin\Debug\dnxcore50\win7-x64\" --configuration \"Debug\" \"C:\temp\dotnet_test\test\"" exited with 1 in 267 ms.
```

Note that we only go down this code path when the command is not on the path, if it is on the path then the command works with the following command line:

```
< dotnet-compile --framework "DNXCore,Version=v5.0" --output "C:\temp\dotnet_test\test\bin\Debug\dnxcore50\win7-x64" --c
onfiguration "Debug" "C:\temp\dotnet_test\test" exited with 0 in 2575 ms.
```

I would submit a PR to remove the escaping but I wanted to understand if there are any cases where it will be necessary. @davidfowl @anurse any history on why this was added?

**Comment by krwq** [comment]
@weshaggard, I'm not sure I understand this issue, could you. IMO the problem is that in CMD you should be escaping quotes with ^ and not with \
When I copy&paste your command I get exactly what you get but when I run:
```
C:\Windows\system32\cmd.exe /C "dotnet-compile --framework ^"DNXCore,Version=v5.0^" --output ^"C:\temp\dotnet_test\test"
```
It works well. For the exception itself - I think we should improve the message

**Comment by weshaggard** [comment]
@krwq the cmd command line is being produced by the tooling which is what the bug is about. All I'm doing is calling `.\dotnet -v publish ..\test -f dnxcore50 -r win7-x64`

**Comment by anurse** [comment]
Sounds like we should change the escaping to use `^`. Does that work when launching processes **without** cmd as well? (i.e. as raw input to `CreateProcess`?)

**Comment by krwq** [comment]
I think it will only work with ShellExecute on Windows although I haven't tested it yet.

**Comment by anurse** [comment]
Hmm. We can't use `ShellExecute`, because it will break output forwarding.

**Comment by krwq** [comment]
I will test it and see. There probably is some easy way to do that

**Comment by brthor** [comment]
When I run like this it also seems to work:

`C:\Windows\system32\cmd.exe /C "dotnet-compile --framework "DNXCore,Version=v5.0" --output "C:\temp\dotnet_test\test""`

Do the quotes need to be escaped?

**Comment by anurse** [comment]
I think they would if the framework had spaces in it.

**Comment by krwq** [comment]
does it ever do?

**Comment by brthor** [comment]
We can test this with spaces in the output param.

`C:\Windows\system32\cmd.exe /C "dotnet-compile --framework "DNXCore,Version=v5.0" --output "C:\temp\dotnet test\test""`

I can't actually get anything compiled for other reasons, but this successfully creates the output folder `dotnet test`

**Comment by anurse** [comment]
Framework doesn't, but certainly some arguments will have spaces, especially paths on Windows.

**Comment by piotrpMSFT** [comment]
We need to make sure we handle platform-specific escaping. "" vs. ^", etc. We're not bringing new standards along with us.

**Comment by krwq** [comment]
Seems like we don't have any specific APIs for platform independent escaping. I'm trying to get us that:
https://github.com/dotnet/corefx/issues/4720

I already found some workarounds (windows: http://www.robvanderwoude.com/escapechars.php OR https://msdn.microsoft.com/en-us/library/xwy0k9bb.aspx ; Linux: http://linux.die.net/man/3/execvp) although that's gonna end up with pretty much creating portable shell execute which we should avoid

**Comment by krwq** [comment]
@brthor I think it is actually purely accidentally working in your case. I believe cmd treats quotes in pairs if you don't escape them with ^. So if you put a string like "a"b"c"  it will be treated as "a" + "b" + "c" (quotes around "a", then non quoted b and then "c")

**Comment by brthor** [comment]
@krwq I am not so sure.

From the `cmd /?`

```
If /C or /K is specified, then the remainder of the command line after
the switch is processed as a command line, where the following logic is
used to process quote (") characters:

    1.  If all of the following conditions are met, then quote characters
        on the command line are preserved:

        - no /S switch
        - exactly two quote characters
        - no special characters between the two quote characters,
          where special is one of: &<>()@^|
        - there are one or more whitespace characters between the
          two quote characters
        - the string between the two quote characters is the name
          of an executable file.

    2.  Otherwise, old behavior is to see if the first character is
        a quote character and if so, strip the leading character and
        remove the last quote character on the command line, preserving
        any text after the last quote character.
```

If I'm understanding that right, because we have more than two quote characters, the outside ones will be stripped and the others will just be passed along as regular quotes, which seems to be what we want in this case.

**Comment by brthor** [comment]
@krwq or are you saying quotes inside of quotes aren't parsed properly without escaping?

**Comment by jasonwilliams200OK** [comment]
@brettfo, @krwq, I wanted to share couple of edge cases maybe related to this work.

The following example is written in JavaScript for node.js REPL, but the underlying problem is of windows cmd. It is simply spawning a `.bat` file with arguments. The catch is the path to bat file has a withespace (and the cases do not fail when path doesn't have whitespaces):

```javascript
// first make sure to enter a directory with space in it.
// second create a file `node2.bat` in this directory and paste
// the following line of code (without slashes):
//
// node %*
//
// then enter REPL with `node` [enter]
// and paste lines one by one

var spawnSync = require('child_process').spawnSync;

spawnSync('node2.bat', ["-p", "'hello'"], { stdio: 'inherit' });
// works
// prints hello as expected.

spawnSync('c:\\users\\jason williams\\node2.bat', ["-p", "'hello'"], { stdio: 'inherit' });
// works
// prints hello as expected

spawnSync('c:\\users\\jason williams\\node2.bat', ['-p', '"hello"'], { stdio: 'inherit' });
// unexpected error:
// 'c:\users\jason' is not recognized as an internal or external command,
// operable program or batch file.

spawnSync('node2.bat', ["-p", "'^'"], { stdio: 'inherit' });
// arguments are mangled
// prints nothing, while '^' was expected

spawnSync('node2.bat', ["-p", '"^^"'], { stdio: 'inherit' });
// arguments are mangled
// prints nothing, while '^^' was expected

spawnSync('node2.bat', ["-p", '"^^^^"'], { stdio: 'inherit' });
// arguments are mangled
// prints '^', while '^^^^' was expected

spawnSync('node2.bat', ["-p", '2', '>', '3'], { stdio: 'inherit' });
// arguments are mangled
// expected to print 'false'
// what actually happens is a file named '3' is created.
```

There are many such corner cases where cmd fails miserably.

**Comment by blackdwarf** [comment]
@Sridhar-MS can you verify and test?

**Comment by krwq** [comment]
@brthor is working on similar issue and this should go away when he merges his work. Please assign back if there is more issues than what you're working on right now.

**Comment by danroth27** [comment]
https://github.com/dotnet/cli/issues/525

**Comment by brthor** [comment]
closed via #985
